### ISSUE TITLE
Hey! Thanks for the great project! I have a question on contributing...

### ISSUE THREAD

**Issue by xpl** [rr-debugger/rr#2773]
The project is awesome. Unfortunately, it crashes when I try to replay the recorded dump of our system (we have all sorts of syscalls, I/O, shared memory, parallelism and stuff). I wish I could debug the problem and contribute to the project, but I lack the knowledge of the Linux kernel stuff needed to do that.

What books/blogs/otherwise literature on Linux system programming you could suggest to make sense of the internals of rr-debugger for newcomers?

**Comment by khuey** [comment]
Can you start by describing how it crashes and what the error message is?

The architecture is described at https://arxiv.org/pdf/1705.05937.pdf. What we do for each individual syscall/etc is based on how that syscall/etc works reading man pages/kernel source where necessary.

**Comment by xpl** [comment]
The error was:

```
`SYSCALLBUF_FLUSH': expecting tracee signal or trap, but instead at `fork'
```

Thanks for pointing out the paper, btw! And what about the general things on how the kernel works, how ptrace works and all other things necessary beyond manpages? :) Assume that I already know C/C++ and how Linux works from the "user" perspective — but I don't know much about the kernel side.

**Comment by khuey** [comment]
That error message indicates a "divergence", where replay didn't match the recording. We missed some sort of non-deterministic input to the program. Unfortunately these sorts of errors are quite difficult to debug.

What revision of rr are you using? And what CPU do you have? Is there any sort of virtualization involved? Would you be willing to share this trace or does it contain private code/data?

"How ptrace works" is pretty difficult to explain. It's exceedingly complex and very poorly documented. I also don't think much understanding of it will be required here beyond that it's used to intercept and record syscalls and signals (in other words the program inputs) for the tracee and to examine the tracee's state when necessary.

**Comment by xpl** [comment]
> Would you be willing to share this trace or does it contain private code/data?

Yep, it contains private data and this is why I need to figure how to debug those issues myself. I also somewhat figured out that it is because of a mismatch and this is something that (in my limited understanding) requires a really deep dive into the codebase to debug it.

I ran it with `rr version 5.4.0`, Ubuntu `18.04.5 LTS (Bionic)`, kernel `5.4.0-48-generic`, under Docker. There was no virtualization, it is a bare metal Linux workstation. CPU is `Intel(R) Core(TM) i7-7800X CPU @ 3.50GHz`

**Comment by khuey** [comment]
Ok. Is this reproducible (in other words if you record the program doing the "same" thing multiple times do you get similar errors)? If it is, a recording made with `record -n` (without syscall buffering) will be easier to debug.

**Comment by GitMensch** [comment]
@xpl Can you share an update - and possibly edit the topic title to something along "chrash on replay (didn't match the recording)"?

**Comment by xpl** [comment]
Feel free to close the ticket.

**Comment by GitMensch** [comment]
The ticket should stay open if you can still reproduce the issue - can you?
